src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,564 @@
+            // the uncore frequency in the efficient - performant range.
+            // A more robust/future proof solution may be to directly query uncore
+            // counters that indicate utilization (when/if available).
+            // For now only L3 bandwith metric is used.
+            double uncore_req = m_resolved_f_uncore_efficient + f_uncore_range * scalability_uncore;
+
+            // Clip uncore request within policy limits
+            if (uncore_req > m_resolved_f_uncore_max || uncore_req < m_resolved_f_uncore_efficient) {
+                ++m_uncore_frequency_clipped;
+            }
+            uncore_req = std::max(m_resolved_f_uncore_efficient, uncore_req);
+            uncore_req = std::min(m_resolved_f_uncore_max, uncore_req);
+            uncore_freq_request.push_back(uncore_req);
+        }
+
+        // Per core freq
+        std::vector<double> core_freq_request;
+        m_resolved_f_core_efficient = in_policy[M_POLICY_CPU_FREQ_EFFICIENT];
+        m_resolved_f_core_max = in_policy[M_POLICY_CPU_FREQ_MAX];
+        double f_core_range = in_policy[M_POLICY_CPU_FREQ_MAX] - in_policy[M_POLICY_CPU_FREQ_EFFICIENT];
+
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            //////////////////////////////////
+            // Core Scalability Measurement //
+            //////////////////////////////////
+            double scalability = (double) m_core_scal.at(domain_idx).value;
+            if (std::isnan(scalability)) {
+                scalability = 1.0;
+            }
+
+            double core_req = m_resolved_f_core_efficient + f_core_range * scalability;
+
+            // Clip core request within policy limits
+            if (core_req > m_resolved_f_core_max || core_req < m_resolved_f_core_efficient) {
+                ++m_core_frequency_clipped;
+            }
+
+            core_req = std::max(in_policy[M_POLICY_CPU_FREQ_EFFICIENT], core_req);
+            core_req = std::min(in_policy[M_POLICY_CPU_FREQ_MAX], core_req);
+            core_freq_request.push_back(core_req);
+        }
+
+        // Set per core controls
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            if(std::isnan(core_freq_request.at(domain_idx))) {
+                core_freq_request.at(domain_idx) = in_policy[M_POLICY_CPU_FREQ_MAX];
+            }
+            if (core_freq_request.at(domain_idx) !=
+                m_core_freq_control.at(domain_idx).last_setting) {
+                // Adjust
+                m_platform_io.adjust(m_core_freq_control.at(domain_idx).batch_idx,
+                                     core_freq_request.at(domain_idx));
+
+                // Save the value for future comparison
+                m_core_freq_control.at(domain_idx).last_setting = core_freq_request.at(domain_idx);
+                ++m_core_frequency_requests;
+                m_do_write_batch = true;
+            }
+        }
```

Comment:
Though I guess that would only work if the governor runs on a system that natively uses core domain for frequency controls. Otherwise, all of the M_NUM_CORE loops here need to change to ask the governor which  `FrequencyGovernor::frequency_domain_type()` to use, or we need to tell the governor which domain we want (latter option doesn't currently exist, but is in one of my upcoming PRs for the balancer agent).

Reply:
From our offline discussion, let's defer this because it also depends on other work:

 - Optional: #2642 would enable you to use FrequencyGovernor without modifying your M_NUM_CORE loops to ask the governor what to use.
 - #2643 to resolve this issue, and to resolve the discovery that FrequencyMapAgent also does not use the helper